terraform/test/main_test.go

```diff
@@ -0,0 +1,152 @@
+}
+
+func TestTerraform(t *testing.T) { suite.Run(t, &TerraformSuite{}) }
+
+func (s *TerraformSuite) SetupSuite() {
+	var err error
+	t := s.T()
+
+	logger.Init()
+	logger.Setup(logger.Config{Severity: "debug"})
+
+	// We set such a big timeout because integration.NewFromEnv could start
+	// downloading a Teleport *-bin.tar.gz file which can take a long time.
+	ctx := s.SetContextTimeout(2 * time.Minute)
+
+	s.AuthSetup.SetupSuite()
+	s.AuthSetup.Setup()
+
+	me, err := user.Current()
+	require.NoError(t, err)
+
+	client, err := s.Integration.MakeAdmin(ctx, s.Auth, me.Username+"-ruler@example.com")
+	require.NoError(t, err)
+
+	pong, err := client.Ping(ctx)
+	require.NoError(t, err)
+	teleportFeatures := pong.GetServerFeatures()
+
+	var bootstrap integration.Bootstrap
+
+	unrestricted := []string{"list", "create", "read", "update", "delete"}
+	role, err := bootstrap.AddRole("terraform", types.RoleSpecV4{
+		Allow: types.RoleConditions{
+			DatabaseLabels: types.Labels(map[string]utils.Strings{"*": []string{"*"}}),
+			AppLabels:      types.Labels(map[string]utils.Strings{"*": []string{"*"}}),
+			Rules: []types.Rule{
+				types.NewRule("token", unrestricted),
+				types.NewRule("role", unrestricted),
+				types.NewRule("user", unrestricted),
+				types.NewRule("cluster_auth_preference", unrestricted),
+				types.NewRule("cluster_networking_config", unrestricted),
+				types.NewRule("session_recording_config", unrestricted),
+				types.NewRule("db", unrestricted),
+				types.NewRule("app", unrestricted),
+				types.NewRule("github", unrestricted),
+				types.NewRule("oidc", unrestricted),
+				types.NewRule("saml", unrestricted),
+			},
+			Logins: []string{me.Username},
+		},
+	})
+	require.NoError(t, err)
+
+	user, err := bootstrap.AddUserWithRoles("terraform", role.GetName())
+	require.NoError(t, err)
+	s.plugin = user.GetName()
+
+	err = s.Integration.Bootstrap(ctx, s.Auth, bootstrap.Resources())
+	require.NoError(t, err)
+
+	identityPath, err := s.Integration.Sign(ctx, s.Auth, s.plugin)
+	require.NoError(t, err)
+
+	s.client, err = s.Integration.NewClient(ctx, s.Auth, s.plugin)
+	require.NoError(t, err)
+
+	s.teleportConfig.Addr = s.Auth.AuthAddr().String()
+	s.teleportConfig.Identity = identityPath
+	s.teleportFeatures = teleportFeatures
+
+	s.terraformConfig = `
+		provider "teleport" {
+			addr = "` + s.teleportConfig.Addr + `"
+			identity_file_path = "` + s.teleportConfig.Identity + `"
+		}
+	`
+
+	s.terraformProvider = provider.Provider()
+	s.terraformProviders = make(map[string]func() (*schema.Provider, error))
+	s.terraformProviders["teleport"] = func() (*schema.Provider, error) {
+		// Terraform configures provider on every test step, but does not clean up previous one, which produces
+		// to "too many open files" at some point.
+		//
+		// With this statement we try to forcefully close previously opened client, which stays cached in
+		// the provider variable.
+		s.closeClient()
```

Comment:
Is this needed when you already close the client at the end of each test?

Also, is this only an issue due to the way `resource.Test` works? Or can this also affect a running terraform provider in production?

Reply:
Yes, this is mandatory. And yes, this due to the way `resource.Test` works. Every test consists of a sequence of steps. Terraform test framework initialises the provider instance at the beginning of every step without cleaning up a previous one. For some reason, there is no cleanup handler in the Terraform test framework so this piece of code is the simple workaround.

`s.closeClient()` at the end of each test, in fact, closes the latest Terraform client created on the very last step of the whole suite. I think, I need to remove these calls and close the latest connection in the suite shutdown handler instead.

It does not affect production. In production Terraform runs from scratch every time.

Reply:
Possibly stupid question: how does closing the client here force Terraform to clean up its Provider set?

Reply:
It does not. Unfortunately, there is no straightforward way to clean up a provider before it quits (https://github.com/hashicorp/terraform-provider-vsphere/issues/618). Let me explain this a bit more detailed to ensure that we are on the same page. 

Terraform provider is the struct which contains resources schema, methods and so on. The most important piece is `ConfigureContext` function, which opens the connection to Teleport and saves it to provider object metadata. It is called by Terraform when it initialises the provider set. The problem is that there is no `Cleanup` or `Close` counterpart. When Terraform finishes it's job in production, it just quits.

Every test consist of the following structures:

```go
Steps: []resource.TestStep{
	{
		Config: create,
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(name, "kind", "cluster_auth_preference"),
			resource.TestCheckResourceAttr(name, "metadata.0.labels.example", "yes"),
			resource.TestCheckResourceAttr(name, "spec.0.disconnect_expired_cert", "true"),
		),
	},
	{
		Config:   create, // Check that there is no state drift
		PlanOnly: true,
	},
	{
		Config: update,
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(name, "kind", "cluster_auth_preference"),
			resource.TestCheckResourceAttr(name, "metadata.0.labels.example", "yes"),
			resource.TestCheckResourceAttr(name, "spec.0.disconnect_expired_cert", "false"),
		),
	},
	{
		Config:   update, // Check that there is no state drift
		PlanOnly: true,
	},
},
```
At the beginning of every step (an array item above), the fresh new instance of the provider gets created and Terraform calls `ConfigureContext`. An old instance stays in memory with the old connection open. So, there would be 4 open connections to Teleport after this test is finished.

At some point, when there are too many open connections, it returns "too many open files" error. I guess, the reason for this error is go grpc client architecture, which is quite complicated and heavy, but I am not 100% sure. There might be other reasons for this. So, what I do here - I close an old provider connection every time Terraform test framework requests a new provider instance.

That would be much better if I could have shared the same connection between all provider instances, but this would require artificial modifications of the `ConfigureContext` function, which I do not want in production. Actually, to solve this issue, I simply need some kind of a connection pool, but I am not sure that provider should be responsible for this.

Reply:
Thanks for the detailed reply.